### ISSUE TITLE
Service Fabric nodes stuck in Disabled, Disabling, Failed or Failed (running) modes

### ISSUE THREAD

**Issue by HaakonL** [Azure/service-fabric-issues#877]
We're on 6.1.456.9494 and the cluster is unusable. I have several nodes that are stuck for many days now, and nothing I do make them work again. 

We are getting quite worried about launching our game APIs on Service Fabric, as there seems to be almost impossible to replace faulted nodes, scale up/down etc without deleting the whole cluster and starting from scratch. For a critical production API this is of course not something we can rely on. EC2 has a great user interface and scales up/down in a few minutes, I really miss the same level of quality in Azure Service Fabric.

We need a more reliable setup before the end of June, what are the plans/roadmap for Service Fabric, the explorer and the azure portal to become a full fledged dev ops solution, where scaling up/down, replacing of faulty nodes and spinning up new instances can be done as easily and elegant as in EC2?

And how do I fix the faulted nodes, please? I tried removing them in the azure portal, tried adding new nodes, tried scaling up and down, tried to reactive the disabled node and a hundred other different things, but they're just stuck...

![cluster](https://user-images.githubusercontent.com/508432/36639862-73ec0b26-1a15-11e8-8dad-1446de32de96.PNG)
![portal](https://user-images.githubusercontent.com/508432/36639863-740621c8-1a15-11e8-9773-38c72aebf5bb.PNG)
![scale](https://user-images.githubusercontent.com/508432/36639879-e055e192-1a15-11e8-8b20-1118ae21f6bc.PNG)

**Comment by ChackDan** [comment]
Thankyou for reporting this @HaakonL and for the screen shots. Sorry about the issues you have hit.  We will need to review the cluster logs to really figure out why you have a node stuck in disabled for that long. To debug, I will need to know the sequence of steps you took before and after.  I agree with you that cluster management operations should not result in you needing to create the cluster. In fact the expectation is that a customer should **not** have to ever recreate a production cluster, if they have followed the guidance we have provided. Can you send me an email at chackdan@Microsoft.com ?  We can set up a short call with you, early this week so that we can review your cluster design, capacity plan, the expectation on the scale out/in, in addition to a quick review of your application design, so that you are set up to go live with confidence. 

it looks like you have a thread with @oanapl  as well, so adding her here as well.

**Comment by pkcsf** [comment]
@HaakonL  To investigate this issues, please open a support ticket from Azure portal for faster resolution with the complete postmortem. Please open a ticket usin https://portal.azure.com/?#blade/Microsoft_Azure_Support/HelpAndSupportBlade  

Please refer https://azure.microsoft.com/en-us/support/options/  for details on support options

**Comment by HaakonL** [comment]
Hi, I already opened a ticket almost a week ago, but no one responds, even though there is a < 8 hour guarantee. It is very frustrating that the cluster is unstable and unpredictable, and then no one responds from the Azure team when we submit a ticket. How can we trust this in a live production setup?

In case anyone from the Azure support team sees this, the ticket # is 118022217692167.

**Comment by ChackDan** [comment]
@pkcsf  - could you please followup and see what happened to this ticket, if possible, let us use that to get this issue to closure.
@HaakonL  - got your email, will respond on the design reviews.

**Comment by pkcsf** [comment]
@HaakonL, I have  looked into the current state the cluster, and I can see that one node **_D1v2_3** is currently struck in to the disabling state, and my further investigation reveals that, this cluster is configured as a Silver **reliability**, and for Silver reliable cluster, it is expected to have  at least 5 node aka primary nodes should be up and running at any point of time to have data integrity.  
In this scenario, I have also noticed that cluster is configured as a Silver **Durability**, so cluster is  giving  protection from any possible data loss due to node removal, so it not allowing the  node removal  operation (Please find some referencing article at the bottom of this reply to get more details about reliability and durability characteristic with some best practice on scaling operation ) 

To mitigate this issue, you have two option .

1.	Add an additional Node to Cluster, sot that new node will be converted into part of primary node, which will eventually allow the _D1v2_3 to remove from cluster.
2.	Change the reliability level from Silver (5 Node)  to Bronze (3 Node) (You can leverage Update-AzureRmServiceFabricReliability CmdLet for this operation)

**please note that recommended solution is 1.** 

Reference : 
https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-capacity  
https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-scale-up-down

**Comment by pkcsf** [comment]
@HaakonL , my fault,  recommended mitigation  **step is 2,** i.e change the reliability level to Bronze by **Update-AzureRmServiceFabricReliability** .

**Comment by HaakonL** [comment]
Hi,

I would rather not change the reliability level downwards at any point -
what is the point with a reliability level then? I find your suggestion
very weird. Fix an issue by lowering the reliability, no way :)

Keep in mind that I never wanted to remove any nodes, I just wanted to fix
the cluster that suddenly had a disabled node that was impossible to
reactivate.

--

Best regards,
*Haakon Langaas Lageng*

*Monkeybin AS*
Office: +47 469 36 836 <+4746936386>
Direct: +47 997 90 091 <+4799790091>

http://monkeybin.com
<http://www.monkeybin.com>

2018-02-27 5:11 GMT+01:00 pkcsf <notifications@github.com>:

**Comment by HaakonL** [comment]
Hi,

I downgraded to Bronze reliability, and it took several hours. I checked in the SF Explorer and saw that the disabled node was no longer in the list of nodes, only 3 healthy nodes was showing.

Then I upgraded to Silver reliability, which also took several hours.

Now, 24 hours later, there are still only 3 nodes in the list in the SF Explorer.

This system seems totally unreliable in my opinion. It takes way too long to fix issues, way too long to get help, way too long to add or remove nodes etc. I have a hard time believing anyone serious is using Service Fabric for critical live operations. If you see a spike in traffic during live operations, there is just no way to respond to it, as even the simplest operation takes several hours to complete.

We are stuck with this for a while, so I just have to go with SF for a while until we can move back to Amazon EC2 / OpsWorks, but until then I really need the bullet proof recipies on how to work with a critical live api that will need to handle around 100 000 requests per minute and sudden spikes of traffic.

**Comment by rishirsinha** [comment]
HaakonL,

Did you scale up the VMs before upgrading the reliability tier?

Thanks,
Rishi

**Comment by rishirsinha** [comment]
I agree with you, that there should be easy to use set of steps (ideally with little intervention on your side). Please work with us in improving this experience. One thing I would point out though is that a 3 node primary node type is not a good option for a production cluster. I would recommend minimum 5 nodes.

**Comment by HaakonL** [comment]
Hi,

If you look at this thread, you'll see that we had 5 nodes, and one got disabled because of a failure in Service Fabric itself. The recommended steps was to reduce the reliability to Bronze to get rid of the disabled node. It is very frustrating that you reply to this thread and point your finger at us 3 nodes are not good for production. We of course know that - we never wanted 3 nodes in the first place.

I did not scale up before upgrading the reliability. There is no easy way to scale up, and according to the documentation the recommended way of scaling is by setting auto scale rules. We have an auto scale rule that defines a MINIMUM of 5 nodes, but the cluster still only have 3.

**Comment by francescocristallo** [comment]
I have the same problem since a year. I spend more time in fixing the cluster (usually creating a new one) than working on the actual code. I'm very disappointed, and I think Service Fabric should not have been GA. Last night I had the VM Scale set stuck again just for downscaling the VM using the Autoscale "Night rule" from 14 to 8, I created a new production cluster, routed all the traffic there, to wake up this morning with the production API down due to the rise in morning traffic and the Autoscale stuck AGAIN with this:

<img width="1029" alt="capture" src="https://user-images.githubusercontent.com/9793825/37357764-3d2938b2-26b7-11e8-916f-3465907fc6da.PNG">

And this is a simple Stateless service Asp.net Core 2 with just API. I opened a ticket for this minutes ago.

Last month I was billed 2k dollars in 3 days since the Autoscale service created more than 70 machines with a limit of 25, all of them unusable.

All of this for Stateless APIs, for the Stateful service I'm really worried to work on it since I can't even create the service with more than 50 partitions, it simply does not work, as per Issue opened here this morning. 

This entire journey with SF it's been a nightmare so far, instead of adding Containers, Linux and tens of options, I think the main focus should be first on the base service on Windows without containers (improving Azure UI panel, SF Explorer + autoscale stability) and then adding other features.

After a year of study and development I really want to make it work, but I'm searching for alternatives too being this unreliability also a high cause of stress.

**Comment by rishirsinha** [comment]
@HaakonL 

First of all my apologies for sounding insincere. The limited point I was making was you should scale up your VMSS as soon as possible to avoid more issues. I hope the issue resolved  after scaling out the VM.

I also apologize for the more superficial response. This is the sequence of events as I see it.

1.  You had nodes getting deactivated.
2. Since the nodes were not getting deactivated you tried a bunch of things including scale up and scale down the nodes.
3. The whole cluster got wedged.
4. Pradeep provided the mitigation with the reduction of reliability tier, which got things unwedged.
5. Then you upgraded the reliability level, and no new VMs showed up.

In all this case you had auto scale enabled. and configured with minimum size of 5 and that was ignored? Can you give us details of the auto scale setup to check what is happening there.

@francescocristallo 

Can you send me details of the auto scale setup (resource id, subscription id and other details) to me and pradeep. We would like to look at why Autoscale is not honoring the constraints.

**Comment by francescocristallo** [comment]
@HaakonL 
I deleted the clusters creating another one and I'm trying to limit the autoscale since every other day the cluster become stuck and I have to create another new. I'll send you the info as soon as it happens again.

Basically while autoscaling (especially down) one or two nodes become stuck and the entire scale system goes crazy, creating and deleting machines that never go up but are counted in the billing system. Note that I'm autoscaling a VMSS with a Stateless Service on it.

I'm waiting for the update 6.1.472 to be available to live clusters since I see in the bugfixes specific:

**Fixed a bug where a node could show up as Invalid even after being removed from the cluster.** 

 Maybe this could fix it. I'll let you know here or open another issue.
I really hope SF will become faster in scaling in the future.

Thanks

**Comment by HaakonL** [comment]
Hi,

we are moving away from Service Fabric, Microsoft is currently totally incapable of creating a robust and trustworthy micro service orchestration system. Just look at all the serious tickets/issues here, this product is a huge danger zone for any company running a mission critical application. As it stands now, we are not convinced that data loss will not occur in our reliable stateful services when scaling up and down happens several times per day.

**Comment by rishirsinha** [comment]
@HaakonL 

Sad to see a customer get frustrated with the platform and leave. Best of luck with whatever infrastructure you go with.

Thanks,
Rishi

**Comment by brutaldev** [comment]
After similar issues experienced today we will also be leaving Service Fabric. Issues like this happen with our services often and today the entire cluster is not accessible and all applications and services have stopped. We cannot afford continuous downtime like this and constantly recreating the cluster from scratch because MS cannot figure out why it system services break. When the system services break inter-service comms fall over, scaling fails, response times slow down to a crawl. In hindsight, we should never have moved away from Cloud Workers. I agree with @HaakonL that Service Fabric is not mature enough yet to rely on for a production system and that poses a major risk to organizations that rely on infrastructure being reliable.

**Comment by newpowera** [comment]
Cannot believe there are so many others with similar experiences. Apparently the current SF is still something that should be in the lab only. Really surprised Microsoft made it GA with such a poor user experience. We also have to rebuild cluster. How can we have a production system running in SF? Disappointed.

**Comment by wpsmmb** [comment]
Unbeliveable that this ticket was simply closed after the customer quits.

**Comment by zacharyhumphrey** [comment]
@rishirsinha, why is this being marked as closed as outlined by @wpsmmb? This is a common issue that we are experiencing as well and are constantly stuck redeploying to fix corrupted instances or replacing the nodes as any updates take longer than redeploying a cluster. This is a small 5 node cluster and at times it takes over an hour for any updates to occur. Microsoft really needs to make this product better and listen to your customers, this product has been so frustrating to manage, deploy, and modify. The documentation is also lacking and needs to be more improved for basic concepts and clearer concepts. It seems as though, very basic concepts within container computing do not quite work correctly in Service Fabric and Microsoft constantly is writing overly complicated workarounds that are so confusing and hard to understand. I fear there is serious lack of understanding of customer issues and concerns related to a product that a majority of the community doesn't feel is in an adequate state for Generally Available.. Also, why can we still not use backup/restore? This is a basic function of any IT department for risk/mitigation, using an ARM template is not quite suitable; because the ability to replicate and existing environment can be very critical when deploying and testing in lower environments to mitigate regression testing risks when deploying newer versions of applications.

**Comment by wacsintegra** [comment]
@rishirsinha, why is this being marked as closed as outlined by @wpsmmb? This is a common issue that we are experiencing as well and are constantly stuck redeploying to fix corrupted instances or replacing the nodes as any updates take longer than redeploying a cluster. This is a small 5 node cluster and at times it takes over an hour for any updates to occur. Microsoft really needs to make this product better and listen to your customers, this product has been so frustrating to manage, deploy, and modify. The documentation is also lacking and needs to be more improved for basic concepts and clearer concepts. It seems as though, very basic concepts within container computing do not quite work correctly in Service Fabric and Microsoft constantly is writing overly complicated workarounds that are so confusing and hard to understand. I fear there is serious lack of understanding of customer issues and concerns related to a product that a majority of the community doesn't feel is in an adequate state for Generally Available.. Also, why can we still not use backup/restore? This is a basic function of any IT department for risk/mitigation, using an ARM template is not quite suitable; because the ability to replicate and existing environment can be very critical when deploying and testing in lower environments to mitigate regression testing risks when deploying newer versions of applications.

**Comment by VipulM-MSFT** [comment]
We sincerely apologize for the frustration caused by prematurely closing this issue. We do understand that the experience of auto-scale or manually scaling the cluster is not ideal and we are looking at ways to improve it.  If you do encounter the issue, please open the support ticket and let us know on this thread. We will use this thread to collect the feedback and shape the solution. Thank you again for being Service Fabric and Microsoft customer and giving us  chance to improve.

**Comment by wacsintegra** [comment]
@VipulM-MSFT is there a better engagement model that can be leveraged for this product? In most cases, companies do not have time to wait days to hear back from Microsoft Support; which seems to be extremely limited in abilities. In my case, it takes a solid one-two days to get any kind of response, which in 99% of the time is an automated response on assignment to a resource as a part of your workflow and work will not even begin on the ticket for 3+ days in most occassions. 

At this point, its much more time effective to start over than wait on Microsoft. Because our businesses do not hault to the errors and we need to get fixes in place today and not 6 months from now. As we have seen it can take up to day for updates to happen and days to get a any real response from support. This is a container based product advertised for an understanding of rapid deployment for scaling to use demands that is *expensive*. The main issues we experience are related to the random corruption that happens without really any good feedback from the errors that are occurring. I had a node just this morning get stuck as "disabled" and refused any actions on the VMSS and just returned "ErrorCode: InternalExecutionError ErrorMessage: An internal execution error occurred." and we had to delete the whole dang cluster, then redeploy our cluster over a secondary node failure that would not present any real errors. 

Also, what is being done to actually leverage the feedback forumn? There appears to be a lot of solid ideas that can help this product, but alas no one from Microsoft has even responded. But from a customer side it appears nothing is happening.

**Comment by wacsintegra** [comment]
@VipulM-MSFT Please address questions that are tagged to your team....

**Comment by wacsintegra** [comment]
@VipulM-MSFT thank you for the response and insights as well as the personel to reach out to for issues when normal support channels are not appropriately being resolve related to the clusters. Do you have any ETA for the updates for providing more documentation so we can be successful with this product?

**Comment by pkcsf** [comment]
Please add pkc@microsoft.com  also with the support ticket number along with chackdan / vipulm

**Comment by VipulM-MSFT** [comment]
We are working on the testing the topology for the stateless workloads with schedule based auto-scale as the first scenario. Additionally, we plan to provide more details on the inner workings of the VMSS and Service Fabric integration to highlight the class of issues you can run in to and guidance to avoid them as well as mitigate them; while we work on improving the integration channel.

**Comment by wacsintegra** [comment]
@VipulM-MSFT Thank you sir, any ETA for getting this information out there?

**Comment by VipulM-MSFT** [comment]
As I mentioned we are working on it actively. We will provide the update when we have an update; in meanwhile if you do have issues, feel free to reach out to us and we can provide specific guidance based on your cluster topology and workload.

**Comment by pksorensen** [comment]
Feel free to see if you can get any information out of this one:
```
"/subscriptions/4c6c1ea4-3605-4411-a44e-b65382627f5b/resourcegroups/sf-gateway-unix/providers/Microsoft.ServiceFabric/clusters/sf-gateway-unix"
```

Its had 5 nodes stuck (other than the 5 primary nodes) for a day or so now. Autoscale based on CPU.

I also share the concern that I dont know if i would trust SF/VMSS to run a mission critial system right now that require scaling - i seen so many errors that require so much manual effort - and then also the time/effort required to help you guys figure out why it not working. 

I can let you know when i see something failing, i can give you permissions to poke around all you wont - i just dont have time to assist - so take what you need from it if it helps you guys make a better product.

**Comment by VipulM-MSFT** [comment]
@pksorensen - thanks which region?

**Comment by pksorensen** [comment]
@VipulM-MSFT WestEurope

It resolved itself now and are back to 5 nodes.

(clicked restart on a disabling node in explorer)

I will write again when it happens again

**Comment by pksorensen** [comment]
@VipulM-MSFT 

Two nodes got stuck in disabling state again. Restarting them in explorer helps some times. 

Since i cant wait for you guys when it happens - are there anything i can do to help. (its ubuntu, and if yo provide me with a script that i should run on the nodes that get stuck, i will do this.) Thinking that the script will dump and upload what ever logs/dumps that would be helpful for you guys.

And I off cause cant claim that its a problem with service fabric, could be application specific - but how would we know that. Thinking if you guys can run something like azure sql on it which must be using some autoscaling - then the rest of us should properly also be able to get it working.

**Comment by VipulM-MSFT** [comment]
@pksorensen - yes I looked at the cluster. The cluster has stateful workloads and the primary node type is being scaled. Do you want to scale both stateful and stateless workload or only stateless workloads?

I am giving a little bit longer answer to help understand the issues.

When a node containing stateful replicas is scaled back in (or being removed), the VMSS controller first asks the Service Fabric for the approval. This allows Service Fabric to create new replica and transfer the data to the new replica. It needs to gracefully shutdown the existing replicas and often times the ChangeRole/Close calls in the application service gets stuck. The system does not take automatic action on dropping / closing the stateful replicas. When you restart the node, the process is restarted and replicas can move the scale-in operation can complete. 

Scaling the primary node type also has other changes, because Service Fabric system services are stateful and constrained to be on the primary node type. During the scale-out operation not only user workload gets scaled, the system services replicas move to balance the replica counts across nodes. During scale-in, the data for the system services also needs to be moved. The data movements and stuck close/change role are the majority reason why the scaling operations take time. 


Following is the guidance you can try to see if that helps with the situation; I know that it will increase the minimum node count to 10 and we are working to reduce it.

- Do not scale primary node type (keep them at 5)
- Add secondary node type (with Silver) and do not scale them below 5
- You can have your workload unconstrained or constrained only to secondary node types 
- If the disabled gets stuck due to close/change role getting stuck, the node can be restarted if there are enough replicas and there is no risk of data loss (we are working on the app that can do this and eventually move the logic in the product)
- Scaling stateful does take time, so be careful when doing CPU based scale that it does not happen too often

Hope this helps. Thank you again for your support.

**Comment by pksorensen** [comment]
Thanks for the suggestions. I will read it again, but i really am stuck at the "10 nodes". This is already way to expensive to run with 5 nodes for something you dont know will take off :) 

Its not on purpose i scale the primary node, i just did not want to have more nodetypes - but maybe i need that. It really pains me that I pay for 5 nodes running at <5% CPU on the primary nodes to have 5 other idling nodes until workloads increase. Maybe there is no way around that and I have to find out how i want to continue from there.  

I have in the past have success with running a second node type at 0 instances and first scale it out when workload is demanded. I am thinking i could do the same here, keep the primary at 5 with no autoscale. Also add the services that do the workload to the primary node, but when scaling is needed do it on the second node type. Then I would have the same setup as today with 5 nodes in idle mode.

I started out with 3 nodes, but found that to have SF/VMSS clean out nodes itself it had to be silver and therefore 5 nodes. So if above works, i could go back to 3 nodes i guess.

**Comment by VipulM-MSFT** [comment]
Agree and that's what we are working on. 

On the first set of guidance:
For the stateless workload, with the right settings we should be able to 3 on the primary node type and with additional work (which we are doing) on the secondary node you can go down to zero. 

For the stateful workload, with the right setting initially, we should be able to do to 3 on the primary node type and 5 on the secondary node type. 


Ideally, you should be able to scale primary node types with minimum as 5 (the setup you have), but we have seen many issues with the apps getting stuck or mistakes being made to remove/re-image one of the original 5 nodes (that includes the seed node). 

I would suggest that for now, continue to use the architecture you have and once we have the automated app that you can tune for your environment to take mitigation action, it should provide a better experience than what you have right now.

**Comment by pksorensen** [comment]
Thanks for the update.

I will stick with what i have and manual fix when it get stuck. 

To make sure i understand it correct . When a node get stuck, what would be the best action to take. (I can recover from state loose at this point in my application)

1. Restart Node (in explorer)
2. Restart Machine (in azure portal)
3. Reimage Machine ( in azure portal)


And another element that is blocking me right now after i started working on scaling is my own cleanup code and i tried to explain what i am doing here: https://stackoverflow.com/questions/52898018/using-service-filter-events-with-registerservicenotificationfilterasync-in-servi (if you guys have anyone who can share some light on that, it would be much appreciated).

**Comment by pksorensen** [comment]
@VipulM-MSFT Can i do a 

<PlacementConstraints>(NodeTypeName==BackEnd)</PlacementConstraints>

that target seed nodes without knowing the NodeTypeName?

**Comment by VipulM-MSFT** [comment]
@pksorensen - I did not understand the question about the placement constraints fully. Feel free to send me an email. 

One possible solution if you do not care about the state and want to scale like stateless, you can:
- Leave the primary node type as is without scaling
- Create another node type with Bronze durability 
- Set the [ReplicaRestartWaitDuration ](https://docs.microsoft.com/en-us/dotnet/api/system.fabric.description.statefulservicedescription.replicarestartwaitduration?view=azure-dotnet#System_Fabric_Description_StatefulServiceDescription_ReplicaRestartWaitDuration) and [QuorumLossWaitDuration ](https://docs.microsoft.com/en-us/dotnet/api/system.fabric.description.statefulservicedescription.quorumlosswaitduration?view=azure-dotnet#System_Fabric_Description_StatefulServiceDescription_QuorumLossWaitDuration)for stateful services to a smaller number (this way they will not get stuck in the quorum loss, but will continue after possbile data loss situation)
- Deploy the app to remove the down nodes of the secondary node type if they are down for more than some reasonable time (this will clean up the scaled-in nodes from the cluster)

**Comment by pksorensen** [comment]
Thanks @VipulM-MSFT - but i think what I ended up doing is fine for now.

I added placement constraints on all statefull services
```
((NodeTypeName==nt1vm) && (NodeName==_nt1vm_0 || NodeName==_nt1vm_1 || NodeName==_nt1vm_2 || NodeName==_nt1vm_3 || NodeName==_nt1vm_4))
```

so when it scales out due to work load, then its only the processing services (those that demand resources) and hopefully thats enough for now. Have not had issues sofar (but also first have time in weekend to test it fully)

I will let you guys know if it do not work :)

**Comment by brianstringfellow** [comment]
Any update from Microsoft on improving how the service fabric handles nodes with problems?

We're running on version 6.4.622.9590 with 'silver' durability and still had to manually intervene to bring a node back. I expect the service fabric and the VMSS to automatically manage bringing back a healthy node. In our case, an application had a memory leak, exhausted the VMs available memory, and the services on that VM failed. The VM remained in a failed state for two days before I manually dealt with it.

**Comment by ChackDan** [comment]
Today on an SF Cluster, we do not proactively do anything when a VM is pegged at 99% CPU or is nearly out of memory, since from our control plane we cannot tell with certainity if something is wrong or not.    We expect the customer to write a machine local watchdog and manage the repair actions.  or put resource governance on the containers or on the job objects, such that VMs do not become unresponsive. 
Having said that , do suggest ways we can conclude with 100% certainty that  we need to initiate a VM move/reboot/reimage.  We want to help and relieve this toil on the customers.

**Comment by aljo-microsoft** [comment]
Closing as all issues have been replied to or customer has indicated they have a solution.
